### ISSUE TITLE
The Types.ObjectId type doesn't include a self reference in the _id field

### ISSUE THREAD

**Issue by BioSurienDG** [Automattic/mongoose#13473]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

7.2.2

### Node.js version

16.18.1

### MongoDB server version

5.0.13

### Description

From: `/types/types.d.ts` - https://github.com/Automattic/mongoose/blob/master/types/types.d.ts#L82-L84
```Typescript
   class ObjectId extends mongodb.ObjectId {
      _id: this;
    }
```

Thus, the `_id` should provide a `this` reference to itself on the objectId fields on the mongoose documents but instead the ObjectId types are just the `mongodb.ObjectId ` versions and don't include the `_id` reference that the mongoose ones are supposed to have

### Steps to Reproduce

```Typescript
export interface ITest {
    title: string;
}

const TestSchema = new Schema<ITest>({
       title: {
            type: String,
            required: true,
        },
}

const Test = model<ITest>(TEST_COLLECTION_NAME, TestSchema);

// Create a test doc

//Find the Test doc
const testDoc = await Test.findOne({})
// testDoc._id is the mongodb.ObjectId type but not the mongoose type of Types.ObjectId (missing _id which is supposed to be included in the mongoose type)
// testDoc._id._id is undefined even though the type definition implies a "this" reference should be on the _id field of the objectId
```

### Expected Behavior

`_id` should be on the Type.ObjectId fields to match the type defn.
```
   class ObjectId extends mongodb.ObjectId {
      _id: this;
    }
```
From: `/types/types.d.ts` - https://github.com/Automattic/mongoose/blob/master/types/types.d.ts#L82-L84

From the example in the steps to reproduce section there should be a `this` reference on the objectId fields

**Comment by BioSurienDG** [comment]
So, I retried this using node 19 since I thought maybe the issue was we were on different node versions but I still have the same issue (see Loom).

I am using the mongoose in-memory server for the tests which should be 5.0.13. 
https://github.com/nodkz/mongodb-memory-server#configuring-which-mongod-binary-to-use
<img width="989" alt="image" src="https://github.com/Automattic/mongoose/assets/98058801/c2952dd7-8707-4588-b5c6-ff3ad468de86">


https://www.loom.com/share/9fc7c0fafdda4182a462c799118182b2

**Comment by BioSurienDG** [comment]
So I updated my mongodb driver 5.6 from 5.5. and that seems to have fixed it so maybe the issue was on their side thanks for your help 🙂

**Comment by vkarpov15** [comment]
Ah that would explain it. If you list mongodb in your `package.json`, make sure the mongodb version lines up _exactly_ with the one Mongoose uses.

But keep in mind: you don't need to list mongodb as a top level dep if you're using Mongoose. You should just be able to import mongodb.

**Comment by hasezoey** [comment]
> You should just be able to import mongodb.

yes that *would* work, but typescript (or something else like a bundler) will complain if not directly installed, if i remember correctly

though it can (without complaints) be imported like `import { mongo } from "mongoose"` thanks to `Mongoose.prototype.mongo = require('mongodb')`

https://github.com/Automattic/mongoose/blob/19896c74984f1061c6097a4fcfc28074bb13c641/lib/index.js#LL1186C47-L1186C47

**Comment by BioSurienDG** [comment]
Thank you so much @vkarpov15 & @hasezoey!'
I will use that in the future to keep the versions in sync 🙂

**Comment by BioSurienDG** [comment]
@hasezoey What happens if I need the `mongodb` lib for a third-party lib like `connect-mongo`? How can I get the `mongoose` instance of `mongodb` into that lib so I don't get errors like the following:

`Cannot find module 'mongodb' from '../../node_modules/connect-mongo/build/main/lib/MongoStore.js'`

**Comment by hasezoey** [comment]
> What happens if I need the mongodb lib for a third-party lib like connect-mongo? How can I get the mongoose instance of mongodb into that lib so I don't get errors like the following:

depends on how that package requires the instance, is it a normal `require('mongodb')` inside the package in question with a `optionalDependency` / `peerDependency`? then it will depend on how the package-manager resolves packages and the most reliable way would be to manually install `mongodb`, only keep in mind to keep it in-sync with the mongoose package version (it will only get updated in major and minor versions of mongoose)
if it is via something like instancing, then the following should work `const instance = package({ mongodb: mongoose.mongo })`

in the case of `peerDependencies` it looks like it wants mongodb via a `peerDependency`

as for your error, i dont know how to resolve it without knowing the context of the error, is it a runtime error, a typescript error, a bundler error or a package manager error?

**Comment by BioSurienDG** [comment]
It is a runtime error and I fixed it as you suggested by installing `mongodb` (I was hoping there was another so that mongoose could manage the version for me but doesn't seem to be).

Thanks so much for all your help 🙂

**Comment by BioSurienDG** [comment]
@hasezoey So, I found if the `mongodb` version I install in package.json matches the version in `mongoose` peer dependency the `_id` field is broken. (i.e. since I upgraded to mongoose v7.3 it is broken again with `mongodb` v5.6 installed (since mongoose upgraded the internal version to 5.6) but if I downgrade to 5.5. it works)

**Comment by vkarpov15** [comment]
I'm not sure how npm handles missing peer dependency entries in `package-lock`.

Is there any way you can provide a standalone project that demonstrates the issue you're seeing? There's a lot of important details here that we have no way of reasonably getting, and without a standalone project we're unlikely to be able to repro this issue.

**Comment by sagrawal31** [comment]
I just upgraded from Mongoose 6 to 7 and started to face this challenge. I don't have `mongodb` listed in my `package.json`.

For me, this is happening for referenced schemas. For example:

```javascript
const bookSchema = new Schema({
  title: 'String',
  author: { type: 'ObjectId', ref: 'Person' }
}

const Book = mongoose.model('Book', bookSchema);
const doc = await Book.findOne()
```

### Before

In mongoose 6, `doc.author._id` was available.

### After

In mongoose 7, that is `undefined`.

*Note:* I'm using `NestJS + Mongoose` and here is my real schema:

```typescript
@Schema({collection: 'engagement', timestamps: timestamps, versionKey: 'v'})
export class Engagement extends Experience {

    @OneToMany(App, {required: true, immutable: true})
    app: App;

    @Prop()
    expireAt: Date;
```

Where `OneToMany` is-

```typescript
/**
 * A DRY decorator to use references in Schema. This uses {@link Prop}.
 * @param schema The schema class to reference.
 * @param options Other options to pass to {@link Prop}.
 * @constructor
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export const OneToMany = (schema: any, options: PropOptions = {}): PropertyDecorator => Prop({
    type: MongooseSchema.Types.ObjectId,
    ref: schema.name,
    ...options as Record<string, any>,
});
```

**Comment by sagrawal31** [comment]
Apologies, I just tested this on a fresh new NestJS project, and I'm not able to reproduce the above. I'll dig more.

**Comment by sagrawal31** [comment]
I tried to dig deeper, tried to uninstall `mongodb-memory-server` too, but same problem exists. Not sure what's the problem.

**Comment by hasezoey** [comment]
@sagrawal31 maybe try to remove the lockfile and the `node_modules` folder and install everything again. 

- is `MongooseSchema` imported from mongoose directly (`import { Schema as MongooseSchema } from "mongoose"`)?
- what nestjs plugin are you using (`@nestjs/mongoose@10.0.0`)?
- where / how are you using the ObjcetId property and is the type at that point?

**Comment by sagrawal31** [comment]
I'm very 😞  @hasezoey that despite reading this entire thread multiple times, it didn't click me to remove `node_modules` altogether. Though I tried to delete the lock file before, but deleting `node_modules` worked. Thanks a ton 🙏🏻

Anyways, to answer your questions-

> is MongooseSchema imported from mongoose directly (`import { Schema as MongooseSchema } from "mongoose"`)?

Yes. `import mongoose, {Schema as MongooseSchema} from 'mongoose';`

> what nestjs plugin are you using (`@nestjs/mongoose@10.0.0`)?

Latest `@nestjs/mongoose@10.0.0`.

> where / how are you using the ObjcetId property and is the type at that point?

I was using it to compare like this- `engagement.app._id.equals(currentApp._id)`.

**Comment by BioSurienDG** [comment]
@vkarpov15 So, I did some testing and I figured out the conflict. Turns out the conflict is between 3 dependencies (bson, mongodb, and mongoose).

I managed to resolve the issue by removing `bson` and just accessing it through `mongoose` since I didn't need it to be installed like I needed with `mongodb`. Not sure if you're interested in fixing this issue but in case you do here is the repo you requested: https://github.com/BioSurienDG/mongoose-bug

If you don't need it anymore lmk and I with delete it.

### Package.json:
```JSON
{
  "name": "mongoose-bug",
  "version": "1.0.0",
  "description": "Testing mongoose _id bug",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "mongodb": "5.6.0",
    "mongoose": "7.3.1",
    "bson": "5.2.0"
  },
  "devDependencies": {
    "typescript": "4.8.4",
    "mongodb-memory-server": "8.13.0"
  }
}
```

**Comment by hasezoey** [comment]
> Not sure if you're interested in fixing this issue but in case you are here is the repo you requested: https://github.com/BioSurienDG/mongoose-bug

from my quick glance at the `package-lock.json`, you seem to manually request `5.2.0` exactly, whereas the installed mongodb  requests `^5.3.0` (at least 5.3.0), meaning you have 2 versions of bson in use, one version is patched by mongoose and the version you directly use is not (i also dont think it is in scope for mongoose to change that)

TL;DR: you have 2 versions installed, where only one is patched by mongoose

**Comment by BioSurienDG** [comment]
Ah that makes sense - confirming that when upgrading to `^5.3.0` it works. I guess basically if I need `bson` installed separately the solution would just be to keep it in sync with `mongoose` (like we did with `mongodb`).

I think we can close this issue then. 

@vkarpov15 @hasezoey Also, it might be good to list the versions of the required libraries in the docs in case someone needs to install them separately?

**Comment by hasezoey** [comment]
> Also, it might be good to list the versions of the required libraries in the docs in case someone needs to install them separately?

those versions will change every version and can be easily be found by accessing the package.json either via a git tag or some site like unpkg.com

also, mongodb is accesible through `mongoose.mongo` and bson is accessible through `mongoose.mongo.BSON`;

**Comment by vkarpov15** [comment]
@BioSurienDG I took a closer look and you're right, the issue is that you need to be absolutely sure that your bson version lines up exactly with the one that Mongoose installs if you choose to list bson in "dependencies". Otherwise you'll have two separate copies of bson. Better to use `mongoose.mongo.BSON`, or just `import bson from 'bson'` without listing bson as a top-level dependency.
